core.d.ts

```diff
@@ -0,0 +1,145 @@
+}
+
+interface ResourceObj<T> {
+	[index: string]: T
+}
+
+type FilterOperationKey = '$ne' | '$eq' | '$gt' | '$ge' | '$lt' | '$le' | '$add' | '$sub' | '$mul' | '$div' | '$mod'
+type FilterOperationValue = Filter
+type FilterOperations = {
+	[P in FilterOperationKey]: FilterOperationValue
+}
+type FilterFunctionKey = '$contains' | '$endswith' | '$startswith' | '$length' | '$indexof' | '$substring' | '$tolower' | '$toupper' | '$trim' | '$concat' | '$year' | '$month' | '$day' | '$hour' | '$minute' | '$second' | '$fractionalseconds' | '$date' | '$time' | '$totaloffsetminutes' | '$now' | '$maxdatetime' | '$mindatetime' | '$totalseconds' | '$round' | '$floor' | '$ceiling' | '$isof' | '$cast'
+type FilterFunctionValue = Filter
+type FilterFunctions = {
+	[P in FilterFunctionKey]: FilterFunctionValue
+}
+
+interface ResourceFilter extends ResourceObj<Filter> {}
+
+type FilterObj = FilterOperations & FilterFunctions & {
+	$raw?: RawFilter
+
+	$?: string | string[]
+
+	$and?: Filter
+	$or?: Filter
+
+	$in?: Filter
+
+	$not?: Filter
+
+	$any?: Lambda
+	$all?: Lambda
+} & ResourceFilter
+
+interface FilterArray extends Array<Filter> {}
+type FilterString = string
+// Strictly speaking `[ string, ...Filter ]` but there isn't a way to type that yet
+type RawFilter = string | (string | Filter)[] | {
+	$string: string
+	[index: string]: Filter
+}
+type Lambda = {
+	$alias: string
+	$expr: Filter
+}
+type Filter = number | FilterString | FilterObj | FilterArray
+
+interface ResourceExpand extends ResourceObj<Expand> {}
+type ExpandObject = ODataOptions & ResourceExpand
+
+interface ExpandArray extends Array<Expand> {}
+type Expand = string | ExpandArray | ExpandObject
+
+type ODataOptions = {
+	$filter?: Filter
+	$expand?: Expand
+	$orderby?: any
+	$top?: any
+	$skip?: any
+	$select?: any
+}
+type OptionsObject = ODataOptions & {
+	[index: string]: string | string[]
+}
+
+type ODataMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
+
+type AnyObject = {
+	[index: string]: any
+}
+
+type ResourceId = string | number | Date
+
+type SharedParam = 'apiPrefix' | 'passthrough' | 'passthroughByMethod'
+type Params = {
+	apiPrefix?: string
+	method?: ODataMethod
+	resource?: string
+	id?: ResourceId
+	url?: string
+	body?: any
+	passthrough?: AnyObject
+	passthroughByMethod?: {
+		GET: AnyObject
+		POST: AnyObject
+		PATCH: AnyObject
+		DELETE: AnyObject
+	}
+	customOptions?: AnyObject
+	options?: ODataOptions
+}
+
+export abstract class PinejsClientCore<T> {
+	apiPrefix: string
+	passthrough: AnyObject
+	passthroughByMethod: AnyObject
+	backendParams: AnyObject
+
+	// `backendParams` must be used by a backend for any additional parameters it may have.
+	constructor(params: string | Params, backendParams: AnyObject)
+
+
+	// `backendParams` must be used by a backend for any additional parameters it may have.
+	clone(params: string | Params, backendParams: AnyObject): T
+
+	query(params: Params): Promise<any>
+
+	get(params: Params): Promise<any>
```

Comment:
I've typed these more accurately for the case of `get`/`query` where we actually perform some validation on the result to know we've got an expected result, and changed the others to use `{}`